core/src/main/java/io/grpc/Metadata.java

```diff
@@ -399,11 +416,47 @@ private String toStringInternal() {
     private final String name;
     private final byte[] asciiName;
 
+    private static BitSet generateValidTChars() {
+      BitSet valid  = new BitSet(0x7f);
+      valid.set('-');
+      valid.set('_');
+      for (char c = '0'; c <= '9'; c++) {
+        valid.set(c);
+      }
+      // Only validates after normalization, so we exclude uppercase.
+      for (char c = 'a'; c <= 'z'; c++) {
+        valid.set(c);
+      }
+      return valid;
+    }
+
+    private static String validateAndNormalizeName(String n) {
+      checkNotNull(n);
+      char[] token = n.toCharArray();
+      checkArgument(token.length != 0, "token must have at least 1 tchar");
+
+      for (int i = 0; i < token.length; i++) {
+        char tChar = token[i];
+        // TODO(notcarl): remove this hack once pseudo headers are properly handled
+        if (tChar == ':' && i == 0) {
+          continue;
+        }
+
+        checkArgument(tChar < VALID_T_CHARS.length() && VALID_T_CHARS.get(tChar),
```

Comment:
Hmm.   You are correct, but now I can't find why I put that in there.  I recall something about index out of bounds, but looking again it actually works.

Done.